src/gui/share.h

```diff
@@ -0,0 +1,205 @@
+                       const QUrl& url,
+                       const QDate& expireDate,
+                       QObject *parent = NULL);
+
+    /*
+     * Get the share link
+     */
+    const QUrl getLink();
+
+    /*
+     * Set a share to be public upload
+     * This function can only be called on link shares
+     *
+     * On success the publicUploadSet signal is emitted
+     * In case of a server error the serverError signal is emitted.
+     */
+    void setPublicUpload(bool publicUpload);
+    
+    /*
+     * Set the password
+     *
+     * On success the passwordSet signal is emitted
+     * In case of a server error the serverError signal is emitted.
+     */
+    void setPassword(const QString& password);
+
+    /*
+     * Is the password set?
+     */
+    bool isPasswordSet();
+
+    /*
+     * Get the expiration date
+     */
+    const QDate getExpireDate();
+
+    /*
+     * Set the expiration date
+     *
+     * On success the expireDateSet signal is emitted
+     * In case of a server error the serverError signal is emitted.
+     */
+    void setExpireDate(const QDate& expireDate);
+
+signals:
+    void expireDateSet();
+    void publicUploadSet();
+    void passwordSet();
+
+private slots:
+    void slotPasswordSet(const QVariantMap &reply, const QVariant &value);
+    void slotPublicUploadSet(const QVariantMap &reply, const QVariant &value);
+    void slotExpireDateSet(const QVariantMap &reply, const QVariant &value);
+
+private:
+    bool _passwordSet;
+    QDate _expireDate;
+    QUrl _url;
+};
+
+/**
+ * The share manager allows for creating, retrieving and deletion
+ * of shares. It abstracts away from the OCS Share API, all the usages
+ * shares should talk to this manager and not use OCS Share Job directly
+ */
+class ShareManager : public QObject {
+    Q_OBJECT
+public:
+    explicit ShareManager(AccountPtr _account, QObject *parent = NULL);
+
+    /**
+     * Tell the manager to create a link share
+     *
+     * @param path The path of the linkshare relative to the user folder on the server
+     * @param password The password of the share
+     *
+     * On success the signal linkShareCreated is emitted
+     * For older server the linkShareRequiresPassword signal is emitted when it seems appropiate
+     * In case of a server error the serverError signal is emitted
+     */
+    void createLinkShare(const QString& path,
+                         const QString& password="");
+
+    /**
+     * Fetch all the shares for path
+     *
+     * @param path The path to get the shares for relative to the users folder on the server
+     *
+     * On success the sharesFetched signal is emitted
+     * In case of a server error the serverError signal is emitted
+     */
+    void fetchShares(const QString& path);
+
+signals:
+    void linkShareCreated(const QSharedPointer<LinkShare> share);
+    void linkShareRequiresPassword();
+    void sharesFetched(QList<QSharedPointer<Share>>);
```

Comment:
should be a const &

Also, QVector is more efficient than QList (but that should not matter much)